nav2_map_server/src/occ_grid_loader.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2008, Willow Garage, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * This file contains helper functions for loading images as maps.
+ *
+ * Author: Brian Gerkey
+ */
+#include "nav2_map_server/occ_grid_loader.hpp"
+
+#include <string>
+#include <vector>
+#include <memory>
+#include <stdexcept>
+#include <chrono>
+
+#include "LinearMath/btQuaternion.h"
+#include "SDL/SDL_image.h"
+
+using namespace std::chrono_literals;
+
+namespace nav2_map_server
+{
+
+const char * OccGridLoader::frame_id_ = "map";
+const char * OccGridLoader::topic_name_ = "occ_grid";
+const char * OccGridLoader::service_name_ = "occ_grid";
+
+OccGridLoader::OccGridLoader(rclcpp::Node * node)
+: node_(node)
+{
+  std::string mode_str;
+
+  // Get this node's default parameter values, using defaults if not supplied in the YAML file
+  node_->get_parameter_or_set("resolution", resolution_, 0.1);
+  node_->get_parameter_or_set("negate", negate_, 0);
+  node_->get_parameter_or_set("occupied_thresh", occupied_thresh_, 0.65);
+  node_->get_parameter_or_set("free_thresh", free_thresh_, 0.196);
+  node_->get_parameter_or_set("mode", mode_str, std::string("trinary"));
+  node_->get_parameter_or_set("origin", origin_,
```

Comment:
I think the best default for origin is (0,0,0) incase there's a malformed yaml file

Reply:
The origin will no longer be a Node parameter but will reside in the user's map YAML file.

Reply:
this comment it outdated - it was back when there were values here